### PR TITLE
Use beginthreadex, and ensure we use the correct C calling conventions in all callbacks

### LINKED ISSUE #1791: Switch __wt_thread_create to use beginthread on Windows
Feedback from Laurent Dupuis on http://groups.google.com/group/wiredtiger-users:

> A thread in an executable that calls the C run-time library (CRT) should use the _beginthreadex and_endthreadex functions for thread management rather than CreateThread and ExitThread; this requires the use of the multithreaded version of the CRT. If a thread created using CreateThread calls the CRT, the CRT may terminate the process in low-memory conditions.

### PR DESCRIPTION
**Pull request by markbenvenuto**
Fixes #1791

There are a few changes:
1. Use beginthreadex to allocate CRT TLS structures at thread creation instead of deferring their allocation. This is also better for Managed C++.
2. Ensure all APIs use __stdcall and __cdecl as appropriate. By default under x64 MSVC this does not make a difference since __stdcall == __cdecl unlike on x86.
3. Add a build variant to MCI system that ensures we use the correct calling conventions everywhere by forcing the default calling convention to __vectorcall. This ensures any missing calling convention annotations are build failures.

Note: I cannot compile the tests, and python wrapper with the stronger warnings so I only verify calling convention compatibility against wiredtiger itself.

Tests:
Windows with and without /Gv

### CODE REVIEW

**Comment by agorrod** [comment]
@markbenvenuto The thread based part of this change looks good to me. I'm less sure about the cdecl component - the set of functions that have the WT_CDECL macro applied to them seems fairly arbitrary to me.

Are the changes tied together?

File: src/async/async_worker.c [hunk 1]
```diff
@@ -281,7 +282,7 @@ __async_worker_op(WT_SESSION_IMPL *session, WT_ASYNC_OP_IMPL *op,
  * __async_worker --
  *	The async worker threads.
  */
-void *
+WT_THREAD_RET /* Quiet style.py */
```
Comment (michaelcahill) [inline src/async/async_worker.c]: What does this mean?

**Comment by markbenvenuto** [comment]
The _beginthreadex function uses __stdcall calling convention, and once I started cleaning up that, the next step was cleanup all the other cases of calling convention mismatches by ensuring the code compiles when __cdecl/__stdcall is not the default calling convention.

The standard C functions [bsearch](https://msdn.microsoft.com/en-us/library/w0k41tbs.aspx), and (qsort)[https://msdn.microsoft.com/en-us/library/zes7xw0h.aspx] have a compare function like this signature:

 int ( __cdecl *compare ) (const void *key, const void *datum) 

That compare function requires __cdecl instead of __stdcall, and since I raised the error level, I had to fix these mismatches.

File: src/async/async_worker.c (hunk 1 above)
Comment (markbenvenuto) [inline src/async/async_worker.c]: When I first hit an error in style.py, I thought it was a bug in my change so I made this hack. The real issue was the name of the function in the comment.

I will send a PR tomorrow to fix a bug in style.py related this issue (ie, pointers as return types confuse style.py)

**Comment by agorrod** [comment]
Thanks @markbenvenuto. lgtm - I'll merge.

**Pull request merged**